### ISSUE TITLE
Discussion: 5min+ build time and 122M of dependencies?

### ISSUE THREAD

**Issue by sylvinus** [yeoman/generator-webapp_DEPRECATED#40]
Feel free to close this at will, but this default yeoman generator is really heavy.

It takes more than 5 minutes to install with dependencies on good hardware and the node_modules directory (which, per guidelines, should be commited) weights a whopping 122M.

Would love to have feedback on this!

**Comment by sindresorhus** [comment]
I agree this has some improvement potential.

I guess a lot of that time is used on building optipng/jpegtran which will [soon be gone](https://github.com/yeoman/node-optipng-bin/pull/6) for most users.

The huge amount of deps is the downside of every dep being local to it's parent. The upside is that you don't have to worry about nested dependency conflict and ugly stuff like that. This is just how Node works. We could look into reducing it a bit, but it will always be a lot. HDD space is cheap though, but network is not. We'll look into what's taking the most space.

Any other thoughts on how we could improve this welcome.

**Comment by pwmckenna** [comment]
What do you mean, "per guidelines, should be commited"?

**Comment by sindresorhus** [comment]
@pwmckenna the recommended practice in Node.js is to commit your deps for sites/apps.

**Comment by sindresorhus** [comment]
So, I looked into this and it turns out one module was using 75MB (!!), we're working on fixing that.

Follow this issue: https://github.com/yeoman/yeoman/issues/944?ref=commandbar

**Comment by alejandroiglesias** [comment]
@pwmckenna @sindresorhus not sure. There's some people stating that commiting modules into your repo is a bad practice. It's a good idea to keep repo clean. You can always `npm install --production` on deploy.

**Comment by sindresorhus** [comment]
@alejandrogarciaiglesias then you have the risk of people removing or force publishing their packages or a patch release breaking something. You can use shrinkwrap, but that will not guarantee checksum consistency, which means the first argument still stands.

**Comment by alejandroiglesias** [comment]
I understand. A theoretical solution to that would be using Grunt for building the project with tasks in the following order:
1. Run build tasks
2. Run tests
3. Copy built files into actual deploy/public directory (only after no tests failed)

Of course, the practical safe side is to do what you suggest.
Cheers.

**Comment by passy** [comment]
When I install `generator-webapp` today, I get a 23M `node_modules/`. This is quite an improvement, even though I think there's still room for improvement.

**Comment by addyosmani** [comment]
I wish we had a table somewhere of total size of scaffolded plus installed
deps for all generators

On Wednesday, July 10, 2013, Pascal Hartig wrote:

> When I install generator-webapp today, I get a 23M node_modules/. This is
> quite an improvement, even though I think there's still room for
> improvement.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/yeoman/generator-webapp/issues/40#issuecomment-20778082
> .

## 

Sent from Gmail Mobile

**Comment by sindresorhus** [comment]
@passy yay, my PRs are starting to take effect. We knew it would take some time for it to bubble up to us. Let's review what's in the dep structure that makes up the most of the 23MB.

**Comment by pankajparashar-zz** [comment]
Today I did a fresh installation of `generator-webapp` with no additional dependency (only H5BP and jQuery). Here goes my observations,

Before Build :

Time taken to install - 1 min 50 secs (110 secs)
Total size of the folder - 57.3 MB
Total size of the dependencies (`node_modules/`) - 56.3 MB

This is still quite large! Not sure how @passy managed to get 23M folder size for dependencies? 

After Build :

Running `$ grunt` for the first time throws this warning,

```
Running "mocha:all" (mocha) task
Testing: http://127.0.0.1:9000/index.html

Warning: PhantomJS timed out, possibly due to a missing Mocha run() call. Use --force to continue.
```

and aborts the build process due to warnings. Although running with `--force` solves the problem but this essentially means, I'll miss out on testing.

**Comment by sindresorhus** [comment]
Yeah, we added two more binaries to grunt-contrib-imagemin which increased the size slightly. Although the main problem is that the biggest size cost is from sub-dependencies not in our control, which waste space by including tests and other junk in their packages. I've had multiple crazes where I've submitted a multitude of PRs, but I can't fix the whole npm registry :/

![screen shot 2013-09-08 at 2 21 31 pm](https://f.cloud.github.com/assets/170270/1103062/40b69f54-1881-11e3-9c25-9d424e88c398.png)

Though I don't think it's too bad. You only install it once per project and HDD space is cheap. I'd much rather have the convenience of npm with not having to think about dependency conflicts, than using something like pip or gem.

## 

For the Mocha thing. Can you open a separate ticket?

**Comment by jaseemabid** [comment]
@sindresorhus How did you generate this picture?

**Comment by sindresorhus** [comment]
@jaseemabid screenshot of DaisyDisk
